keps/sig-api-machinery/storage-migration.md

```diff
@@ -0,0 +1,346 @@
+---
+kep-number: 25
+title: Migrating API objects to latest storage version
+authors:
+  - "@xuchao"
+owning-sig: sig-api-machinery
+reviewers:
+  - "@deads2k"
+  - "@lavalamp"
+approvers:
+  - "@deads2k"
+  - "@lavalamp"
+creation-date: 2018-08-06
+last-updated: 2018-08-06
+status: provisional
+---
+
+# Migrating API objects to latest storage version
+
+## Table of Contents
+
+   * [Migrating API objects to latest storage version](#migrating-api-objects-to-latest-storage-version)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+      * [Proposal](#proposal)
+         * [Alpha workflow](#alpha-workflow)
+         * [API](#api)
+         * [Failure recovery](#failure-recovery)
+         * [Beta workflow - Automation](#beta-workflow---automation)
+            * [a. Exposing the configured storage versions via API](#a-exposing-the-configured-storage-versions-via-api)
+            * [b. Using the apiserver version to trigger migration](#b-using-the-apiserver-version-to-trigger-migration)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Alternatives](#alternatives)
+         * [“Active management”](#active-management)
+         * [update-storage-objects.sh](#update-storage-objectssh)
+
+
+## Summary
+
+We propose to extend and improve the [oc adm migrate storage][] command line
+tool to migrate the stored API objects in Kubernetes clusters. We will
+deliver a standalone tool of alpha quality in 2018 Q3. We will integrate the
+storage migration into the Kubernetes upgrade process in 2018 Q4. We will make
+the migration automatically triggered in 2019.
+
+[oc adm migrate stroage]: https://www.mankier.com/1/oc-adm-migrate-storage
+
+## Motivation
+
+"Today it is possible to create API objects (e.g., HPAs) in one version of
+Kubernetes, go through multiple upgrade cycles without touching those objects,
+and eventually arrive at a version of Kubernetes that can’t interpret the stored
+resource and crashes. See k8s.io/pr/52185."[1][]. We propose a tool that
+prevents this problem.
+
+[1]:https://docs.google.com/document/d/1eoS1K40HLMl4zUyw5pnC05dEF3mzFLp5TPEEt4PFvsM
+
+### Goals
+
+A successful storage version migration tool must:
+* work for Kubernetes built-in APIs, custom resources (CR), and aggregated APIs.
```

Comment:
I'm not sure if we should handle aggregated APIs, which can have their own storage conventions. The read-then-write operations might not migrate the stored data. The API can even be non-idempotent.

Reply:
The proposal isn't final yet. 

I want to discuss if storage migrator should take care of aggregated apis. If so, let's decide on the two alternatives listed in the `Beta workflow - Automation` section. Then I'll update the proposal.